### PR TITLE
LighterWriter - write/save predictions

### PR DESCRIPTION
**Pull request by ibro45**
A callback for writing predictions.

I looked into Lightning's BasePredictionWriter. it's a very basic and not really useful Callback, so instead of inheriting from it, I thought that it's cleaner to start from a clean Callback.

Right now, it's only possible to write the predictions according to their dataloader/batch index. Will look into having an optional `get_names()` method in a`Dataset`s, which, if present, would replace the indices as names.

### CODE REVIEW

**Comment by ibro45** [comment]
Right now, `write_as` allows you to specify only one type - but what if `pred` is a list or dict of different types? For example, a model that outputs a segmentation mask and a description of the segmented object.

Maybe extend `write_as` so that it can be:

-  a single type - `pred`'s values, no matter if only one or multiple, will be saved as that type.
- list of types - positionally correspond to each element of a `pred` list.
- dict of types -  correspond to a `pred` dict per key.

**Comment by kbressem** [comment]
I'd like it to be a writer base class and multiple sub-classes, such as `SegmentationWriter`, `ClassificationWriter`, etc. Cramming all into one class, will be to much. Should a model have multiple outputs, we can chain multiple writers.

**Comment by surajpaib** [comment]
I very much agree with @kbressem on this.

I don't know if we want to think about every possible use case independently. I think that's why lightning also provided a base prediction writer.  

We can ship our default writer which works for common use cases, as you have implemented. For more specific use cases like you mentioned, I say let the users handle these, there can be way too much variability. For instance, I save my predictions in a `csv` format, and I would use my custom writer over the lighter writer. I'm sure there would be more such scenarios from users.

**Comment by ibro45** [comment]
Even if we do per-task writers, we still have a problem of mapping them to each output in multi-value `pred`s. `pl.Trainer(callbacks=...)` accepts a single Callback or a list of Callbacks. If it's a list you can't just map them to a `pred` of type `list`, as there may be other callbacks there, like `pl.callbacks.ModelCheckpoint`. Since it doesn't accept a dict, you also can't match it to a `pred` of type `dict`. And trying to fix that would be extremely hacky.

Moreover, I don't think we should split `LighterWriter` per task (`SegmentationWriter`, `ClassificationWriter`, ...). Reasons:

1. This diverges from `LighterSystem`'s unified behavior. There are no `LighterClassificationSystem` or similar. Everything is based on what kind of data is being passed through the model and criterion. Obviously, there are some assumptions, but they're pretty straightforward, and only for some wild-usecase you \*may\* need to pay closer attention.
Similarly, `LighterLogger` unifies multiple loggers and you and select them by just setting the flag for each on or off. I don't think it would've make much sense there to have our own `WandbLogger`, `TensorboardLogger`, when really the only difference between then is log call (`tensorboard.add_scalar(), `wandb.log()`,...)
2. `LighterWriter` itself is pretty simple, separating it out feels unnecessary and a lot of behavior will be shared without adding to much on top of it. E.g., both in `SegmentationWriter` and `ClassificationWriter`, you'd like to also have an option to save the tensor as-is to `.pt` file perhaps. The main thing that would be separated out into different classes would be the [`_write()` method](https://github.com/project-lighter/lighter/blob/68559d6928893337b05eda94ef233963367c20f9/lighter/callbacks/writer.py#L48), and I think it's just too simple and unnecessary to separate. Obviously, it'll be slightly more complex if it supports mapping different `write_as` types to different outputs in `pred`, but that's about it.

Regarding the CSV format, that will [be supported here too](https://github.com/project-lighter/lighter/blob/68559d6928893337b05eda94ef233963367c20f9/lighter/callbacks/writer.py#L22), and that's going to be very simple too (similar to `LighterLogger`). The idea of the `LighterWriter` is to cover as much as possible in the simplest way, so that the user does not have to write their own `Writer` `Callback`. They will still be able to do so if they wish, obviously, as they can with the logger and the system, overriding the existing or writing from scratch.

**Comment by surajpaib** [comment]
Maybe I was a bit muddy in my response and misread what @kbressem mentioned. What I meant to say was we don't need to think of handling every possible case in the `LighterWriter`. Of course, that would be the default writer and handle a bunch of use cases. `lighter` won't have any other `Writer` than the default one, but users can define their own subclasses using the `LighterWriter` to support any specific use cases they might have, and we should try to support that in the cleanest way possible. 

To address your previous comment,
I think having `write_as` mimic the format of output predictions should work very well, i.e, if output is a dict, then `write_as` is also a dict with types. Of course, with some simplifications such as broadcasting types.

> But that is the essence of lighter; we try to provide as much as possible out-of-the-box and, hopefully, in a simple way. Lightning gives you the building blocks, so you have to code it up. We want to eliminate the coding part as much as possible and let it all/most be config-defined.

100% agree, but this also leads us to some complications

I'm still of the opinion that we shouldn't try to handle all use cases in `LighterWriter` because this easily leads to a very large and bloated config with too many flags. 
- If I have multi-output predictions, what if I want to write them in a specific folder per item? which is probably more sensible than having a flat output structure. This will need another flag unless we assume that multi-output will also be folder structured.
- For medical imaging use cases, we want images saved as NRRD/NIFTI. This means more functionality is added to the `LighterWriter` and gets into more specific behavior, which deviates from the more general approach of `lighter`. 
- Users may want arbitrary types in the `write_as`. For example, some may want to save predicted output in a `json` format which is common for a lot of CV tasks such as detection. 

We can ofcourse try to support as many use cases as we think of, but we also need to consider being easily extendable.

**Comment by surajpaib** [comment]
Made some aesthetic comments for now. I will test the PR with some code tomorrow. Overall looks pretty good, seems like you thought of everything that's possible (including 24 FPS video 😉) !

We should add some test cases for such multi-use case features. 
1. It will help the person submitting the PR to test for a range of conditions
2. Reviewers can easily understand all the functionality offered, and so can the users when needed. 

Maybe we make that a requirement from now on? It doesn't have to be something very crafty since people would already have test cases while implementing functionality, it should be easy to have it added to the PR. It will automatically increase coverage as well. 
@ibro45 @kbressem Wdy think?

File: lighter/callbacks/utils.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+    if not is_valid:
+        logger.error(
+            f"`{name}` has to be a Tensor, List[Tensor], Tuple[Tensor],  Dict[str, Tensor], "
+            f"Dict[str, List[Tensor]], or Dict[str, Tuple[Tensor]]. `{type(data)}` is not supported."
+        )
+        sys.exit()
+
+
+def structure_preserving_concatenate(
```
[the remark this answers is not in the document]
Comment (surajpaib) [reply lighter/callbacks/utils.py:93]: From what I understand of this function, I would suggest a rename to `collate` as it seems to operate very similarly to collate functions. Correct me if I'm mistaken here.

**Review by surajpaib — CHANGES_REQUESTED** [review]

File: lighter/utils/model.py [hunk 2]
```diff
@@ -53,14 +56,59 @@ def replace_layer_with_identity(model: Module, layer_name: str) -> Module:
     return replace_layer_with(model, layer_name, Identity())
 
 
-def remove_last_layer_sequentially(model: Module()) -> Sequential:
-    """Removes the last layer of a network and returns it as an nn.Sequential model.
-    Useful when a network is to be used as a backbone of an SSL model.
+def adjust_prefix_and_load_state_dict(model: Module, ckpt_path: str, ckpt_to_model_prefix: Dict[str, str] = None) -> Module:
```
Comment (ibro45) [inline lighter/utils/model.py:59]: Wanted to make this a separate PR, but I somehow ended up pushing it here. Guess I'll just leave it here

File: lighter/callbacks/writer.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        write_as: Optional[Union[str, List[str], Dict[str, str], Dict[str, List[str]]]],
+        write_on: str = "step",
+    ) -> None:
+        self.write_dir = Path(write_dir) / datetime.now().strftime("%Y%m%d_%H%M%S")
+        self.write_as = write_as
+        self.write_on = write_on
+
+        self.parsed_write_as = None
+
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/writer.py:48]: https://docs.monai.io/en/stable/_modules/monai/transforms/transform.html#Transform
https://github.com/Lightning-AI/lightning/search?p=1&q=ABC

I think it's bad practice not stating that something abstract if that's what it is. Also, I think seeing that something is abstract right away is actually readable. `NotImplemented` should be used for stuff that should be implemented in that particular code but it isn't. Abstract, on the other hand, tells you use this blueprint to build your stuff.

File: lighter/callbacks/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/utils.py:93]: lol i might've reinvented collate here - gotta check if it'd work with torch `default_collate`. Need tests for that.

File: lighter/callbacks/writer.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            logger.error("`write_on` must be either 'step' or 'epoch'.")
+            sys.exit()
+
+        # Broadcast the `write_dir` so that all ranks write their predictions there.
+        self.write_dir = trainer.strategy.broadcast(self.write_dir)
+        # Let rank 0 create the `write_dir`.
+        if trainer.is_global_zero:
+            self.write_dir.mkdir(parents=True)
+        # If `write_dir` does not exist, the ranks are not on the same storage.
+        if not self.write_dir.exists():
```
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/writer.py:92]: yes

**Comment by surajpaib** [comment]
@ibro45 Are we ready to merge this?

**Comment by ibro45** [comment]
Writing tests

**Comment by ibro45** [comment]
Alright if you agree let's merge it, so that we can merge other PRs, and I'll make another one for tests
